Port the full MISMIP+ test group from the old vesrion of compass

andrewdnolan opened the pull request:
<!--
Thank you for your pull request.
Please add a description of what is accomplished in the PR here at the top:
-->
__Description__: 
Currently the only test case within the MISMIP+ test group is a `SmokeTest`, which uses an archived "spun-up" initial condition. The goal of the PR is add more (if not all) of the experiments from the MISMIP+ experiments as test cases for MALI ([Asay-Davis et al. 2016](https://gmd.copernicus.org/articles/9/2471/2016/gmd-9-2471-2016.pdf)). 

----
<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
__Checklist__:
* [ ] User's Guide has been updated
* [ ] Developer's Guide has been updated
* [ ] API documentation in the Developer's Guide (`api.rst`) has any new or modified class, method and/or functions listed
* [ ] Documentation has been [built locally](https://mpas-dev.github.io/compass/latest/developers_guide/building_docs.html) and changes look as expected
* [ ] The `E3SM-Project` submodule has been updated with relevant E3SM changes
* [ ] The `MALI-Dev` submodule has been updated with relevant MALI changes
* [ ] Document (in a comment titled `Testing` in this PR) any testing that was used to verify the changes
* [ ] New tests have been added to a test suite

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

trhille reviewed (changes requested):
@andrewdnolan, this is looking good! I made some comments regarding compass conventions and whatnot, but this has come a long way in a short time.  However, I'm getting an error when running this on Perlmutter. Here's what I take to be the important points from the traceback:
```
  File "/global/cfs/cdirs/fanssie/users/trhille/compass/compass/landice/tests/mismipplus/setup_mesh.py", line 265, in _setup_MISMPPlus_IC
    src['dirichletVelocityMask'].loc[:] = xr.where(mask, 1, 0)
    ...
    ValueError: could not broadcast input array from shape (1,1,3456) into shape (1,3456,11)
    ```
It looks like it created the `landice_grid.nc` file successfully, but is throwing an error when trying to populate the mesh.

trhille commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,285 @@
+    # Boolean maks for indexes which correspond to the N/S boudnary of mesh
+    mask = (src.yCell == src.yCell.min()) | (src.yCell == src.yCell.max())
+    # Set the velocity boundary conditions
+    src['dirichletVelocityMask'].loc[:] = xr.where(mask, 1, 0)
```
This is the line that's causing the error. It needs to be
```suggestion
    for lev in range(nVertLevels):
        src['dirichletVelocityMask'].loc[:,:,lev] = xr.where(mask, 1, 0)
```
or something similar.

andrewdnolan replied:
Given that both data arrays in the offending line have the `nCells` coordinate, `xarray` _should_ automatically align the coordinates.  

It looks like this is a bug that was resolved as of xarray version `2023.9.0`. Here's the issue describing the bug and the PR which fixed it: 
  - **Issue**: https://github.com/pydata/xarray/issues/7030
  - **PR**: https://github.com/pydata/xarray/pull/8067 


From the discussion in the issue, it seems like: 
```
src['dirichletVelocityMask'] = xr.where(mask, 1, 0).variable
```
resolves the problem. Maybe we go that route, over the `for` loop since it's a little bit closer to what the code actually should be?

andrewdnolan commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,285 @@
+    # Hard code some parameters from Table 1. of Asay-Davis et al. 2016
+    accum = 0.3       # m^{-1}
+    C = 3.160e6       # Pa m^{-1/3} s^{1/3}
+    rhoi = 918        # kg m^{-3}
```
I'm happy to use 910, but Table 1 from [Asay-Davis et al. 2016](https://gmd.copernicus.org/articles/9/2471/2016/gmd-9-2471-2016.pdf) has 918. So I guess if we were strictly following MISMIP+ we'd use that. I guess the difference would be _really_ small in the actual simulation results. 

Since this test case is for other modelling purposes besides the actual MISMIP+ experiments, should I move it to the config file and set at 910, but add a comment that we are departing from MISMIP+ conventions? 

That way if at some point there was desire to actually run MISMIP+ there wouldn't be need to change any code?

matthewhoffman replied:
I like the idea of making rhoi a cfg option for this particular test case.  It is important to be able to reproduce the prescribed MISMIP+ protocol if needed, but we otherwise prefer not to use pure ice density.  Putting it in a cfg where it can be used to automatically update both the initial condition calculations and the namelist is the safest, most flexible approach, I think.

andrewdnolan commented on compass/landice/tests/mismipplus/__init__.py:
```diff
@@ -1,5 +1,6 @@
-from compass.testgroup import TestGroup
+from compass.landice.tests.mismipplus.makemeshtest import MakeMeshTest
```
Thanks for the help with the compass naming conventions! I've gone with `mesh_gen` for the test case name and renamed the testcase class to `MeshGen`. From looking at other examples that seems to be consistent with conventions, but if not let me know.

matthewhoffman replied:
@andrewdnolan , I'm still seeing `MakeMeshTest` here.  Is the change to `mesh_gen` something you need to push yet?

andrewdnolan commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,285 @@
+        The mesh to be shifted
+    """
+
+    # Why do you need unique grid points?
```
@matthewhoffman  As you mentioned, for the translation `unique_xs.min()` is unnecessary since `xCell.min()` is equivalent, so I've just switched to the simpler option here. 

But, later on when we alter the `yEdge` length on the North/South boundaries  `unique_ys.min()` is still necessary because we need to alter both the boundary and the first interior edge on the North/South boundaries. `yCell.min()` can get us the boundary index but not the first interior, so I've left this approach there.

andrewdnolan commented on compass/landice/tests/mismipplus/makemeshtest/__init__.py:
```diff
@@ -0,0 +1,39 @@
+        resolution : int
+            The distance between horizontal (x-y) grid points
+        """
+        name = "MakeMeshTest"
```
Changed to `"mesh_gen"`.

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,285 @@
+            (ds_mesh.yEdge == unique_ys_edge[-1]))
+    # WHL: zero out the edges on the boundary
+    #      (not necessary because velocity will also be zero)
+    ds_mesh['dvEdge'] = xr.where(mask, 0.0, ds_mesh.dvEdge)
```
I agree this particular manipulation seems extra unusual.  The idea was to effectively eliminate the northern half of the northernmost cells and similar for the southernmost row.  I'm skeptical this is actually affecting anything in the model, but given that we ran MISMIP+ successfully with this, my preference is to retain it.  If we run into complication stemming from it down the road, we can adjust.

andrewdnolan replied:
Sounds good. Seems like the safest course of action to get these MISMIP+ like experiments running.

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+        """
+        # Defined resolutions, the value of `resolution` in the configuration
+        # file needs to a key in the dictionary, or added manually.
+        resolution_params = {'8km': {'nx': 73, 'ny': 14, 'dc': 9237.604307},
```
@trhille , for your usage of this domain for convergence testing with calving on, you will presumably want a gutter beyond the prescribed initial ice shelf calving front to allow for portions of the calving front to advance under prognostic calving without hitting the edge of the mesh.  My recollection is the `nx` values specified here give 2 extra cells (at all resolutions) beyond the MISMIP+ fixed calving front position.   

If allowing a gutter option sounds desirable now, I think that would be a new cfg option, either with units of number of cells or distance.  Do you want that included in this PR so you can use it out of the gate for your planned tests?

trhille replied:
Yes, good idea. I don't have a strong preference for whether it should be specified in number of cells or distance, but it should end up being ~the same distance for all resolutions, so maybe specifying distance is the simplest options.

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+    src = xr.open_dataset(filename)
+
+    # Use `.loc[:]` for setting the initial conditions  since we are setting
+    # the fields with scalar values. This ensures values are properly broadcast
```
I'm not familiar with the `.loc` attribute, but I'm confused by this comment.  It looks the fields below are spatially varying so I don't follow the comment about scalar values here.

andrewdnolan replied:
Yeah I need to update that comment to make it more accurate. I'm using it here to ensure broadcasting and alignment work correctly in variable assignment. When a scalar value is being assigned to a field `.loc[:]` will ensure that the scalar value is broadcast against all the dimensions of the field and that the broadcasting is aligned properly . For example, if we left off the `.loc[:]` and did something like `src.example_field = 910.` then the resulting example field would have all it's dimensions (e.g. `Time`, `nCells`, `nVertInterfaces`) dropped and just be a scalar value. 

But the need for `.loc[:]` in field assignment is more general. For example the `mask` variable in the `_setup_MISMPPlus_IC` function  only contains dimensions `nCells` (since it was created from `src.yCell` which itself only contains dimensions `nCells`). So when we use the `mask` variable to assign fields (e.g. `dirichletVelocityMask`) the inclusion of `.loc[:]` ensure that the value we are assigning are broadcast and aligned against all the dimensions of the field on the right hand side (`Time`, `nCells`, `nVertInterfaces`). 

So more generally, `.loc[:]` is needed anytime any time a variable is being used for assignment which does not contain all the dimensions of the field you're trying to set. Therefore, `.loc[:]` is needed in all the IC prescriptions because all variables in the mesh contain a `Time` dimension, but none of the fields used to set the IC contain that `Time` dimension. 

Hopefully that description is clear. I'll update the comment here to reflect the details as elaborated here more accurately.

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+
+    # Use `.loc[:]` for setting the initial conditions  since we are setting
+    # the fields with scalar values. This ensures values are properly broadcast
+    # against the fields coordinates
```
```suggestion
    # against the field's coordinates
```

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+    # Convert SMB from m/yr to kg/m2/s
+    accum *= rhoi / spy
+    # Set the surface mass balance
+    src['sfcMassBal'].loc[:] = xr.where(src.xCell > xcalve, accum, -100.)
```
The original implementation here of using strong negative SMB to maintain calving position is fine for now.  @trhille , if you use these meshes for evolving calving fronts, you'll want to remove this.  We could replace this method with `calvingMask` instead (which didn't exist when the original MISMIP+ test case was written).  I don't think that needs to happen in this PR, unless you think it would be timesaving to have that implementation switched now.

trhille replied:
Good catch. I think eventually we do want to replace this with `calvingMask`, but I don't care much whether that is done in this PR or in a short follow-up. I guess we might as well do it now, since I think it is an easy change.

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+    # Set the velocity boundary conditions
+    # NOTE: `.variable` is needed so that coordinates are properly broadcast
+    #        due to a bug in xarray, which was resolved as of 2023.9.0
+    src['dirichletVelocityMask'].loc[:] = xr.where(mask, 1, 0).variable
```
Add a comment that the prescribed north and south b.c. are free slip.  This is not a common b.c. we use in MALI and requires a special section in the albany_input.yaml file.  uReconstructX needs to be zero, but that is not explicitly set here because that's the default value when the field is created in the input file.
 
Note: the albany input file for MISMIP+ here:
https://github.com/MPAS-Dev/compass/blob/main/compass/landice/tests/mismipplus/albany_input.yaml
seems to have lost the free slip boundary condition information from the old repo:
[https://github.com/MPAS-Dev/MPAS-Model/blob/landice/develop/testing_and_setup/compass/landice/MISMIP%2B/albany_input.yaml and added to the test grou](https://github.com/MPAS-Dev/MPAS-Model/blob/landice/develop/testing_and_setup/compass/landice/MISMIP%2B/albany_input.yaml#L9-L10)
@mperego , can you advise us how to correct that?

mperego replied:
I think you need to add back
```
    Dirichlet BCs: 
      SDBC on NS dirichlet for DOF U1 prescribe Field: dirichlet_field
```
otherwise Albany will prescribe that Dirichlet BCs to both component of the velocity.

andrewdnolan replied:
Thanks for catching this @matthewhoffman! And thanks for the info @mperego. 

I'll update the `albany_input.yaml` file to reflect this.

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+    # Set the effectivePressure using a Weertman power law
+    src['effectivePressure'].loc[:] = C
+
+    # Set the layerThicknessFractions so that layers are evenly distributed
+    src['layerThicknessFractions'].loc[:] = 1.0 / float(nVertLevels)
```
I don't think there is any reason why this needs to be case (I don't see a specification in the MISMIP+ paper), and if not I would prefer to stick with our default nonuniform vertical spacing.  I'm guessing this was specified in the old test case before we had implemented a standard nonuniform spacing option in our landice mesh creation script.  @trhille, what do you think? (if you interpolate from our archived MISMIP+ 2km state, this will be a difference to be aware of)

trhille replied:
I agree that we don't need to have uniform layer thickness. When interpolating from the archived state, I don't know if we will be interpolating 3D velocity, since the point of this (at least for now) is to use the velocity solver. We could also punt on this for now and have a config setting for 'uniform' or 'glimmer' layer spacing.

andrewdnolan replied:
Great, I'll add a config option. How does `vetical_layer_distribution` with the options `uniform` or `glimmer` sound? I'll leave the default as `glimmer` so the `SetupMesh` step doesn't alter the vertical layer distribution, as created by the `create_landice_grid_from_generic_MPAS_grid.py`, by default.

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+    # Get the center of y-axis (i.e. half distance)
+    yCenter = (ds_mesh.yCell.max() + ds_mesh.yCell.min()) / 2.
+
+    # Shift x-axis so that the x-origin is all the way to the left
+    xShift = -1.0 * ds_mesh.xCell.min()
+    # Shift y-axis so that it's centered about the MISMIP+ bed trough
+    yShift = 4.e4 - yCenter
```
@andrewdnolan , these shifts are slightly different from the old MISMIP+ test case, right?  If I followed things, these shifts are more sensible.  However, @trhille, be aware that if you interpolate the state from our archived MISMIP+ 2km state file, the coordinate system of these new meshes will be slightly different unless you do extra work to align them.

matthewhoffman commented on compass/landice/tests/mismipplus/smoke_test/__init__.py:
```diff
@@ -1,11 +1,13 @@
     """
     A test case for running a smoke test of the MISMIP+ configuration
+
+    This test case  uses a pre-made mesh file.
```
```suggestion
    This test case performs a forward simulation for a short duration
    using a pre-made mesh file with a 2 km MISMIP+
    spin-up that has previously been run to steady state
```

trhille commented on compass/landice/tests/mismipplus/mesh_gen/mesh_gen.cfg:
```diff
@@ -0,0 +1,16 @@
+# ice density (kg m^{-3}). MISMIP+ uses 918 (Table 1. Asay-Davis et al. 2016),
+# but MALI defaults to 910. The user can choose if they want to stricly follow
+# MISMIP+ or use the default MALI value.
+ice_density = 910.
```
Sure, that's fine with me if we are including the option in the config file.

andrewdnolan commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+
+        # unpack the mesh parameters for the given resolution as set as
+        # attributes
+        self.nx, self.ny, self.dc = resolution_params[key].values()
```
Good catch, I didn't necessarily think of this. But it turns out that since python 3.6+ dictionaries are now ordered in "insertion order". [Here's a description](https://docs.python.org/3.6/whatsnew/3.6.html#new-dict-implementation) of the changes that were implemented: 

> The [dict](https://docs.python.org/3.6/library/stdtypes.html#typesmapping) type now uses a “compact” representation based on [a proposal by Raymond Hettinger](https://mail.python.org/pipermail/python-dev/2012-December/123028.html) which was [first implemented by PyPy](https://morepypy.blogspot.com/2015/01/faster-more-memory-efficient-and-more.html). The memory usage of the new [dict()](https://docs.python.org/3.6/library/stdtypes.html#dict) is between 20% and 25% smaller compared to Python 3.5.
> 
> The order-preserving aspect of this new implementation is considered an implementation detail and should not be relied upon (this may change in the future, but it is desired to have this new dict implementation in the language for a few releases before changing the language spec to mandate order-preserving semantics for all current and future Python implementations; this also helps preserve backwards-compatibility with older versions of the language where random iteration order is still in effect, e.g. Python 3.5).

But given that it says the order preserving aspect shouldn't be relied upon, I'll switch to an `OrderedDict` so that I can keep the concise implementation.

andrewdnolan replied:
### Option 1
Well maybe`OrderDict` isn't the easiest way. While it works, it makes it a little harder to read. Here's what would be needed: 
```python
resolution_params = {'8km': OrderedDict([('nx', 73),  ('ny', 14), ('dc', 9237.604307)]),
                     '4km': OrderedDict([('nx', 144), ('ny', 24), ('dc', 4618.802154)]),
                     '2km': OrderedDict([('nx', 284), ('ny', 44), ('dc', 2309.401077)]),
                     '1km': OrderedDict([('nx', 566), ('ny', 84), ('dc', 1154.700538)])}
```
then 
```python
self.nx, self.ny, self.dc = resolution_params[key].values()
```
should work version independent. 

### Option 2
An alternative and easier method as compared to option 1 is: 
```python
mesh_params_dict = resolution_params[key].values()

for resolution_param in mesh_params_dict: 
    setattr(self, resolution_param, mesh_params_dict[mesh_params_dict])
``` 
but still not super easy to read. 

### Option 3
There's always the brute force option, which while not super concise is very legible:
```
mesh_params_dict = resolution_params[key].values()

self.dx = mesh_params_dict['dx']
self.dy = mesh_params_dict['dy']
self.dc = mesh_params_dict['dc']
```
---
While this isn't super important for the PR, which do you prefer?

andrewdnolan commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+        # check if the resolution has been changed since the `compass setup`
+        # command was run
+        if self.resolution != resolution:
+            warnings.warn(f'Resolution was set at {self.resolution:2d}km when'
```
Your right, it probably makes more sense to raise an error. This version of the code raises a warning and then ignores the changes to resolution made after the time `compass setup` was run. But I could definitely see someone getting confused as to why the their resolution change didn't work. The warning will go to the log file, and it would frustrating to have to dig through that to figure out why the resolution change didn't work.

andrewdnolan commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,301 @@
+    # Shift y-axis so that it's centered about the MISMIP+ bed trough
+    yShift = 4.e4 - yCenter
+
+    # Shift all spatial points by calculated shift
+    for loc in ['Cell', 'Edge', 'Vertex']:
+        ds_mesh[f'x{loc}'] = ds_mesh[f'x{loc}'] + xShift
+        ds_mesh[f'y{loc}'] = ds_mesh[f'y{loc}'] + yShift
```
Thanks @matthewhoffman!

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,473 @@
+        ds_mesh = cull(ds_mesh, logger=logger)
+        ds_mesh = convert(ds_mesh, logger=logger)
+
+        # custom function to shift the orgin for the MISMIP+ experiments
```
```suggestion
        # custom function to shift the origin for the MISMIP+ experiments
```

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,473 @@
+        # find the number of y rows (`ny`) from the `nominal_dy`
+        nominal_ny = Ly / nominal_dy
+
+        # find the acutal number of y rows (`ny`) by rounding the `nominal_ny`
```
```suggestion
        # find the actual number of y rows (`ny`) by rounding the `nominal_ny`
```

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,473 @@
+
+    def calculate_dc(Ly, ny):
+        """
+        Calcuate the edge length that conforms to the desired `ny`
```
```suggestion
        Calculate the cell spacing that conforms to the desired `ny`
```

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,473 @@
+            nx = np.ceil(Lx / dc)
+            # The modulo condition below ensures there is exactly one cell
+            # past the the desired domain length. So, when no gutter
+            # infromation is provided, add an extra column to make
```
```suggestion
            # information is provided, add an extra column to make
```

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,473 @@
+            # the default gutter length 2
+            nx += 1
+        else:
+            # ammend the domain length to account for the gutter
```
```suggestion
            # amend the domain length to account for the gutter
```

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,473 @@
+    dc = calculate_dc(Ly, ny)
+    nx = calculate_nx(Lx, dc, gutterLength)
+
+    # add two to `ny` accomodate MISMIP+ specific culling requirments.
```
```suggestion
    # add two to `ny` to accomodate MISMIP+ specific culling requirments.
```

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,473 @@
+    nx = calculate_nx(Lx, dc, gutterLength)
+
+    # add two to `ny` accomodate MISMIP+ specific culling requirments.
+    # This ensures, after the culling, that the cell center too cell center
```
```suggestion
    # This ensures, after the culling, that the cell center to cell center
```

matthewhoffman reviewed (changes requested):
@andrewdnolan , I did a quick review now to get you quick feedback.  Overall, this looks fantastic.  All of my comments are nitpicky things.  I appreciate the level of detail that went in to getting a robust product here, and despite what I said initially about this mesh probably not getting used much, we keep finding new reasons we'll need it in the coming months, so I think this was worth doing right.  I'll review again once the docs are added, and I will then actually test it by running it.

andrewdnolan commented on compass/landice/tests/mismipplus/spin_up/__init__.py:
```diff
@@ -0,0 +1,58 @@
+
+        super().__init__(test_group=test_group, name=name)
+
+        config = CompassConfigParser()
+        module = 'compass.landice.tests.mismipplus.spin_up'
+        # add from config
+        config.add_from_package(module, 'spin_up.cfg')
```
I just wanted to chime in here with a little context as to why this is here. I don't think this section is necessary in most cases, but this was my work around to get the resolution (as passed in the config file) in the directory structure. Adding the config here allows the resolution to be parsed before the step is initialized, and therefore the resolution value can be passed to the `subdir` argument in the `Step` initialization. 

Maybe there's a better way to do this, but this was just what I came up with from looking through the compass repo (with a particular focus on the mesh convergence tests which has resolution in the directory structure).

matthewhoffman replied:
@andrewdnolan , thanks for the explanation - that's really helpful.  So what would happen if the user is specifying a cfg file in `compass setup` that has a different value from the default here?

andrewdnolan replied:
Then an error would be raised. At the time of `compass setup` the `resolution` value is stored as an attribute of the step. Then at compass run, the `resolution` is read from the config again and if the value from the config differs from the stored value an error is raised: 

 https://github.com/andrewdnolan/compass/blob/d0b93c4f0b5fe7a7e701aac0964989bb8d3a6343/compass/landice/tests/mismipplus/setup_mesh.py#L85-L97

The error should only pertain to `resolution` since it's the only one stored as an attribute. Other parameters (e.g. gutter, density) should have no problem if changed at runtime.

matthewhoffman replied:
I guess my question is if as written it is even possible to change things at setup time - if a user supplied a modified cfg file with `-f`, whatever value in there is ignored in preference for the version of the cfg file that is actually sitting in the repo, right?

andrewdnolan replied:
Hmm, I guess I'm not 100% clear by what you mean at "setup time."

For parameters in the config file, excluding `resolution`, it is possible to change values between running `compass setup` and `compass run`. I did some crude testing by printing to the log file parameter values changed between these two steps (e.g. `ice_density=200` from reference of `918`) and can confirm the values are updated accordingly. 

For resolution is it not. While that does make setting up various resolutions a little cumbersome, I think this limitation is needed since the directory structure is only created at the `compass setup` stage. If we didn't apply this restriction we could have a 2000m mesh sitting in the directory named 8000m, for example.

matthewhoffman replied:
Compass supports a 'user' cfg file with `-f` that can contain any cfg sections they want, e.g. specifying the allocation to submit jobs against plus some custom test case settings.  I tried doing that and got incorrect behavior like this:
```
cd compass/landice/tests/mismipplus/spin_up
cp spin_up.cfg blah.cfg
edit blah.cfg: resolution = 4000.
compass setup -n 95 -w $scratch5/compass-mismip+-spinup-blahcfg -f blah.cfg
ls $scratch5/compass-mismip+-spinup-blahcfg/landice/mismipplus/spin_up
```
and I see:
```
8000m  job_script.sh  load_compass_env.sh  spin_up.cfg  test_case.pickle
```
In other words, hardcoding retrieving the cfg from the repo eliminates the ability to use a user-specified cfg file with `-f`.  (I realize now I might not have made the question clear.)  I think you just want to change this bit of code to use self.config, which will already be populated with the merged cfg from the defaults in the repo plus the user one.

andrewdnolan replied:
Thanks for the clarification about the `-f` option to `compass setup`, I wasn't really familiar with it. 

I've revamp the way the directory structure is created, mainly by overloading the `compass.TestCase`'s `configure` method. Doing things this way ensures configuration passed over the command line (via `-f`) are properly resolved and accounted for. 

Would be great for you to experiment and corroborate that the fix works on your end. This overhaul was added in [4b0f35a](https://github.com/MPAS-Dev/compass/pull/737/commits/4b0f35a01e4606b80d1b86e37910645cfd780e92).

andrewdnolan commented on compass/landice/tests/mismipplus/streams.landice:
```diff
@@ -48,10 +44,6 @@
     <var name="xtime"/>
     <var name="daysSinceStart"/>
     <var name="deltat"/>
-    <var name="allowableDtACFL"/>
-    <var name="allowableDtDCFL"/>
-    <var name="calvingCFLdt"/>
-    <var name="dtCalvingCFLratio"/>
```
Okay! Works for me. Are they only needed in the `globalStatsOutput` stream? Or would you like them in the `restart` and/or `output` streams?

matthewhoffman replied:
Just in the globalStats file is good.

andrewdnolan replied:
Added!

andrewdnolan commented on compass/landice/tests/mismipplus/smoke_test/__init__.py:
```diff
@@ -21,11 +25,19 @@ def __init__(self, test_group):
 
-        ntasks = 36
-        min_tasks = 4
+        resolution = '2000m'
```
Fixed! Thanks for catching that.

andrewdnolan commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,473 @@
+                            f' the resolution value in the configuration file'
+                            f' within the python module and rerun the `compass'
+                            f' setup` command in order to create a mesh at a'
+                            f' resolution of {resolution:2d}km')
```
Thanks for catching this! Fixed in commit `519189e`.

matthewhoffman commented on compass/landice/tests/mismipplus/albany_input.yaml:
```diff
@@ -3,6 +3,10 @@
 ANONYMOUS:
   Build Type: Tpetra
 
+  Problem:
```
@mperego , MISMIP+ requires a specified rate factor of "A=6.338 × 10−25 Pa−3 s−1 Glen’s law coefficient = 2.0 × 10−17 Pa−3 a−1", and it looks like that setting has gotten lost at some point.  Can you instruct us the current way to specify that in the yaml file here?

matthewhoffman replied:
@mperego , also do we need this bit?
https://github.com/MPAS-Dev/MPAS-Model/blob/landice/develop/testing_and_setup/compass/landice/MISMIP%2B/albany_input.yaml#L24-L27

mperego replied:
> @mperego , MISMIP+ requires a specified rate factor of "A=6.338 × 10−25 Pa−3 s−1 Glen’s law coefficient = 2.0 × 10−17 Pa−3 a−1", and it looks like that setting has gotten lost at some point. Can you instruct us the current way to specify that in the yaml file here?

You should add this lines in the `Problem` section
  ```
Problem: 
    LandIce Viscosity:
      Type: 'Glen''s Law'
      Flow Rate Type: Uniform
      Glen's Law A: 6.338e-25
```
You can avoid the last line if you set the value in MPAS.

mperego replied:
> @mperego , also do we need this bit? https://github.com/MPAS-Dev/MPAS-Model/blob/landice/develop/testing_and_setup/compass/landice/MISMIP%2B/albany_input.yaml#L24-L27

You don't really need it.

matthewhoffman commented on compass/landice/tests/mismipplus/mismipplus.cfg:
```diff
@@ -0,0 +1,13 @@
+max_cells_per_core = 5000
+
+# number of cells at 1000m resolution (with no gutter present). This is used
+# as huerisitc to scale the number of cells with resolution, in order to
```
```suggestion
# as heuristic to scale the number of cells with resolution, in order to
```

matthewhoffman commented on compass/landice/tests/mismipplus/spin_up/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+class SpinUp(TestCase):
+    """
+    Test case for create the MISMIP+ mesh, initial conditions,
```
```suggestion
    Test case for creating the MISMIP+ mesh, initial conditions,
```

matthewhoffman commented on compass/landice/tests/mismipplus/tasks.py:
```diff
@@ -0,0 +1,96 @@
+    return ntasks, min_tasks
+
+
+def _approx_cell_count(config):
```
Is this function called anywhere?  Maybe I'm missing it somehow, but I don't see a call to it.

andrewdnolan replied:
Oh, thanks for catching that Matt! Your totally right that the function is never called, but it definitely should be. 

The plan was to replace [Lines 34-40](https://github.com/andrewdnolan/compass/blob/bc691ee5a0ed5d1a235953e6db512ed6ea914b22/compass/landice/tests/mismipplus/tasks.py#L34-L40) of that same file with the function call. 

While I flushed out the function, I never actually deleted the old Lines 34-40 and replaced them with the function call. Related to your comment below, I guess that also why the function never returned anything although it definitely should. 

Both of the those things have been fixed and I'll push the related commit in just a second.

matthewhoffman commented on compass/landice/tests/mismipplus/tasks.py:
```diff
@@ -0,0 +1,96 @@
+        new_area = (Lx + gutterLength) * Ly
+        # scale by the approx cell count by the relative area increase
+        # due to presence of the gutter
+        cell_count *= ref_area / new_area
```
Related to previous question, why is there nothing being returned from this function?

matthewhoffman reviewed (changes requested):
Another quick pass with a few changes.  I'll follow up with actually running this in the next round!

trhille commented on compass/landice/tests/mismipplus/spin_up/streams.spin_up:
```diff
@@ -0,0 +1,25 @@
+        clobber_mode="overwrite">
+</stream>
+
+<stream name="globalStatsOutput"
```
It would be helpful to add `<var name="albanyVelocityError"/>` to the globalStats stream, so we can make sure the velocity solver is converging.

andrewdnolan replied:
Thanks for the suggestion here! I added this in [`7ef36a1`](https://github.com/MPAS-Dev/compass/pull/737/commits/7ef36a1d01d58112c6d06979deb4d13d8346b261).

trhille commented on docs/developers_guide/landice/test_groups/mismipplus.rst:
```diff
@@ -4,39 +4,98 @@ mismipplus
+layout of mesh rows is symetric in the y direction. 
+
+The ``resolution`` parameter from the ``[mesh]`` section of the configuration
+file is the nominal resolution, where the actual resolution is itteratively 
```
```suggestion
file is the nominal resolution, where the actual resolution is iteratively 
```

trhille commented on docs/developers_guide/landice/test_groups/mismipplus.rst:
```diff
@@ -4,39 +4,98 @@ mismipplus
+with as many as 5000 cells per core (the default ``max_cells_per_core``).
+
+A complication of using this function is that the number of cells in a mesh
+is not know unitl the :py:class:`compass.landice.tests.mismipplus.setup_mesh.SetupMesh`
```
```suggestion
is not known until the :py:class:`compass.landice.tests.mismipplus.setup_mesh.SetupMesh`
```

trhille commented on docs/developers_guide/landice/test_groups/mismipplus.rst:
```diff
@@ -4,39 +4,98 @@ mismipplus
+is not know unitl the :py:class:`compass.landice.tests.mismipplus.setup_mesh.SetupMesh`
+step has been run (i.e. at ``compass run``) for a given ``resolution``, but we 
+need to know how many cores and nodes we will use at setup 
+(i.e. at ``compass setup ...``) since thats when the job script is created. 
```
```suggestion
(i.e. at ``compass setup ...``) since that is when the job script is created. 
```

trhille reviewed (changes requested):
@andrewdnolan, this is excellent stuff! Thanks for your hard work. I have a handful of requested changes, but nothing major.

matthewhoffman commented on compass/landice/tests/mismipplus/spin_up/namelist.spin_up:
I also like 0.8 - good catch

andrewdnolan replied:
Thanks for the heads up here! Added in [7ef36a1](https://github.com/MPAS-Dev/compass/pull/737/commits/7ef36a1d01d58112c6d06979deb4d13d8346b261).

matthewhoffman commented on compass/landice/tests/mismipplus/mismipplus.cfg:
```diff
@@ -0,0 +1,13 @@
+# few cores are available)
+max_cells_per_core = 5000
+
+# number of cells at 1000m resolution (with no gutter present). This is used
```
Based on how I understand this is being used, I think this could be hardcoded and removed from the cfg file.

andrewdnolan replied:
Your totally right. We actually really don't want the user playing with the number, since it would throw off the `ntasks` calculations. 

I've removed the option and comment, and moved both to the [`tasks.py` file](https://github.com/andrewdnolan/compass/blob/4b0f35a01e4606b80d1b86e37910645cfd780e92/compass/landice/tests/mismipplus/tasks.py#L6-L9).

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,479 @@
+        dc = float(ds_mesh.dcEdge[0])
+
+    # get the distance between edges
+    de = 0.5 * dc * np.sin(np.pi / 3)
```
yeah, this is using periodic_grid meshes only, so we'll always have perfect hexagons and this formula will hold

andrewdnolan replied:
I've added a comment which explains/justifies the equilateral assumption, per you request.

matthewhoffman commented:
@andrewdnolan , when I try to set up the mismip+ smoke test, I get a cfg error:
```
  File "/usr/projects/climate/mhoffman/mambaforge/envs/dev_compass_1.2.0-alpha.9/lib/python3.10/configparser.py", line 794, in get
    raise NoOptionError(option, section)
configparser.NoOptionError: No option 'gutter_length' in section: 'mesh'
```

Looks like we need tweak the cfg hierarchy a bit.

matthewhoffman commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,479 @@
+        # check if the resolution has been changed since the `compass setup`
+        # command was run
+        if self.resolution != resolution:
+            raise Exception(f'Resolution was set at {self.resolution:2d}m'
```
I got this exception to be raised with the `-f` config testing I mentioned above.  In so doing, I got this value error for this line:
```
  File "/usr/projects/climate/mhoffman/compass/mismip+_spinup/compass/landice/tests/mismipplus/setup_mesh.py", line 94, in run
    raise Exception(f'Resolution was set at {self.resolution:2d}m'
ValueError: Unknown format code 'd' for object of type 'float'
```

andrewdnolan replied:
Thanks for testing this! I wasn't really aware of the `-f` option. 

I've revamped the way the directory structure is created (see response to your original comment). With that revamp, this bug should be fixed and configuration options passed from the command line should be properly parsed.  (Also the error you got in the `SmokeTest` should be fixed). 

All that was pushed in [4b0f35a](https://github.com/MPAS-Dev/compass/pull/737/commits/4b0f35a01e4606b80d1b86e37910645cfd780e92).

andrewdnolan commented on compass/landice/tests/mismipplus/spin_up/spin_up.cfg:
```diff
@@ -0,0 +1,29 @@
+# Number of vertical levels
+levels = 10
+
+# Initial ice thickness (m)
```
Added a bit more info to the comment in commit `7ef36a1`. Tried to explain where/why this is needed.

andrewdnolan commented on compass/landice/tests/mismipplus/setup_mesh.py:
```diff
@@ -0,0 +1,479 @@
+        dc : float
+            Edge length [m]
+
+        gutterLength: float
```
I've gone to a consistent snake-case naming scheme as of [7025f22](https://github.com/MPAS-Dev/compass/pull/737/commits/7025f22a8b8540ecded117ac6e9295816dceb2dd). Thanks for the heads up about this!

andrewdnolan commented on compass/landice/tests/mismipplus/spin_up/spin_up.cfg:
```diff
@@ -0,0 +1,29 @@
+# entire domain in the y direction is exactly the required y domain length (80 km).
+resolution = 8000.
+
+# length (m) to extend the eastern domain boundary by. Needed for simulations
```
In the MISMIP+ protocol, the ice front and eastern boundary are both located at 640 km. (See $L_\rm{x}$ and $x_{\rm calve}$ in Table 1. of[ Asay-Davis et al. [2016]](https://gmd.copernicus.org/articles/9/2471/2016/gmd-9-2471-2016.pdf)). 

As for the default, it's out of view here, but it's included in the comment a little further down. The default gutter length is two gridcells downstream of the ice front. Any `gutter_length` less than 2*resolution is ignored in favor of the default value.

trhille commented on compass/landice/tests/mismipplus/run_model.py:
```diff
@@ -56,17 +72,20 @@ def __init__(self, test_case, name='run_model', subdir=None,
         self.suffixes = suffixes
+
+        if resolution is None:
+            resolution = 0.0
```
How does this work?

andrewdnolan replied:
Yeah this is definitely strange but has to do with the very different way `resolution` is set between `SmokeTest` and `SpinUp` test cases. 

## `SmokeTest` 
So, this last round of commits removes the option to specify the `SmokeTest` resolution in a config file. The problem with the former approach is a user could request a resolution not supported by test case (currently, any value other than `2000m`). With the current code, the supported resolutions are added as individual test cases at the test group level:
https://github.com/andrewdnolan/compass/blob/67b26525f73724c1d6ba0b2f5d7c724427a2c9e0/compass/landice/tests/mismipplus/__init__.py#L19-L21
(This also has the added benefit of the supported resolutions being shown in `compass list`). __So__, for the `SmokeTest` a resolution is always passed to the initialization and in the code above `resolution` will never be `None`. 

## `SpinUp` 
Okay, so now we get to why a "dummy" value of `0.0` is specified. For the `SpinUp` we want to support arbitrary resolution, which should be set in the users config file. The big hiccup here is that we want to use the requested resolution in the directory structure, but the config file which isn't fully parsed until after the step is initialized. Ideally we'd be able to pass `resolution` into the initialization of the `SpinUp` testcase, but that doesn't work. (We can parse the resolution value in the compass repo's config before initialization, but can not parse a user specified config file (e.g. `compass setup -f ....`) prior to the testcases initialization). So, instead we have this dummy value of `0.0` and instead set the value of the `resolution` attribute in the `SpinUp.configure` method so that we have access to all the config options: 
https://github.com/andrewdnolan/compass/blob/67b26525f73724c1d6ba0b2f5d7c724427a2c9e0/compass/landice/tests/mismipplus/spin_up/__init__.py#L49-L83

So, in short the `SmokeTest` knows the value of the `resolution` attribute at the time of initialization, whereas the `SpinUp` needs to parse `resolution` from the config options, so a temporary dummy value is used until the `resolution` attribute is set by the `SpinUp.configure` method. 

Now, you might ask: Why set the dummy value of `0.0` at all? Why not just leave the `resolution` attribute unset for the `SpinUp` testcase until the configure method is called? That's definitely an option, but in talking to Xylar that seems like poor python practice. I guess it's best to initialize all attributes within the `__init__` method. 

Happy to ditch this if need be in order to have a bit clearer but less "proper" python code.

trhille replied:
Okay, this is fine. I just wanted to make sure it was actually doing something intentional.

andrewdnolan replied:
I added a comment in [f916f60](https://github.com/MPAS-Dev/compass/pull/737/commits/f916f60ccc9d7b8e515ef6a8a775fd71979c5b40) which summarizes the message above to provide a brief explanation to new/future users.

trhille commented on compass/landice/tests/mismipplus/spin_up/__init__.py:
```diff
@@ -0,0 +1,89 @@
+        ----------
+        test_group : compass.landice.test.mismipplus
+            The test group that this test case belongs to
+
```
```suggestion
```

trhille commented on compass/landice/tests/mismipplus/spin_up/__init__.py:
```diff
@@ -0,0 +1,89 @@
+        test_group : compass.landice.test.mismipplus
+            The test group that this test case belongs to
+
+        """
```
```suggestion
        """

```

trhille commented on compass/landice/tests/mismipplus/spin_up/__init__.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        Set up the directory structure, based on the requested resolution.
+        """
+        # get the config options from the TestCase, which
```
```suggestion

         # get the config options from the TestCase, which
```

trhille commented on compass/landice/tests/mismipplus/spin_up/__init__.py:
```diff
@@ -0,0 +1,89 @@
+
+            # store the resolution (at the time of `compass setup`) as an
+            # attribute. This is needed to prevent the changing of resolution
+            # between `compass setup` and `compas run`, which could result
```
```suggestion
            # between `compass setup` and `compass run`, which could result
```

trhille commented on compass/landice/tests/mismipplus/tasks.py:
```diff
@@ -0,0 +1,114 @@
+import xarray as xr
+
```
```suggestion


```

trhille commented on compass/landice/tests/mismipplus/tasks.py:
```diff
@@ -0,0 +1,114 @@
+
+    cell_count : int
+        Number of horizontal cells in the mesh. The value of this parameter is
+        caclulated (or approximated) using the functions below.
```
```suggestion
        calculated (or approximated) using the functions below.
```

trhille commented on compass/landice/tests/mismipplus/tasks.py:
```diff
@@ -0,0 +1,114 @@
+        The minimum number of tasks
+    """
+
+    # read MPI related parameters from configuration file
```
Can we just use `config.getint()` instead of getting floats and then converting to floats below?

andrewdnolan replied:
I've switched to parsing all the parameters as `int`s and then casting them as floats where need be. This was done as of [f916f60](https://github.com/MPAS-Dev/compass/pull/737/commits/f916f60ccc9d7b8e515ef6a8a775fd71979c5b40). Thanks for catching this!

trhille replied:
Do we need to cast these as floats at all? It does no harm, but it will give exactly the same answer as using ints.

andrewdnolan replied:
Your right, in python casting as floats is kind of overkill. Especially since all of the floating point operations with `int`s we are doing are division, where we are explicitly using the floating point operator (e.g. `/` vs. `//`). 

I was just trying to follow best practice, but seems a little unnecessary here and at the expense of readable code. I'll remove the casting as floats to make things a little more concise.

trhille replied:
Thanks. I wasn't sure what was considered best practice here, but I think it is more readable without casting them as floats.

trhille commented on compass/landice/tests/mismipplus/tasks.py:
```diff
@@ -0,0 +1,114 @@
+    cell_count : int
+        the number of cells in the mesh
+    """
+    # get cell count from mesh
```
```suggestion

     # get cell count from mesh
```

trhille commented on compass/landice/tests/mismipplus/tasks.py:
```diff
@@ -0,0 +1,114 @@
+    cell_count : int
+        the number of cells in the mesh
+    """
+    # Fixed domain lenghts [m] (without gutter)
```
```suggestion

    # Fixed domain lenghts [m] (without gutter)
```

trhille commented on docs/users_guide/landice/test_groups/mismipplus.rst:
```diff
@@ -3,26 +3,93 @@
-More tests may be added later.
-There are no config options.
+The ``mismipplus`` test group (:py:class:`compass/landice/tests/mismipplus`)
+implements a limited number of the expeirments from the Marine Ice Sheet 
```
```suggestion
implements a limited number of the experiments from the Marine Ice Sheet 
```

trhille commented on docs/users_guide/landice/test_groups/mismipplus.rst:
```diff
@@ -3,26 +3,93 @@
+implements a limited number of the expeirments from the Marine Ice Sheet 
+Model Intercomparison Project Plus (MISMIP+; 
+`Asay-Davis et al., 2016 <https://gmd.copernicus.org/articles/9/2471/2016/gmd-9-2471-2016.pdf>`_).
+To date, the spinup and ``Ice0`` are the only expeirments currently supported. 
```
```suggestion
To date, the spinup and ``Ice0`` are the only experiments currently supported. 
```

trhille reviewed (changes requested):
@andrewdnolan, I just have a handful of minor comments. I tested this by creating and running 1-, 2-, 4-, 8-, and 16-km meshes, and everything seems to be functioning as expected!

andrewdnolan commented on compass/landice/tests/mismipplus/spin_up/namelist.spin_up:
```diff
@@ -0,0 +1,5 @@
+config_run_duration = '20000-00-00_00:00:00'
```
Gotcha, I've switched to `config_stop_time` for both the `SmokeTest` and the `SpinUp`.

andrewdnolan commented:
# Testing 
All testing done on `chicoma`
 
## `SmokeTest`

__Setup command__: 
`compass setup -n 95 -w /lustre/scratch5/anolan/mismip+_pr_testing`

__Then__:  
```
cd /lustre/scratch5/anolan/mismip+_pr_testing/landice/mismipplus/smoke_test/2000m
sbatch job_script.sh
```
__Output__: 
![Screenshot 2024-02-21 at 12 43 16 PM](https://github.com/MPAS-Dev/compass/assets/32367657/0d2c937b-6d01-4bc4-be8d-9e3dbd78f26e)


## `SpinUp`
Added some account/job related config, plus all resolution / parameter variations (e.g. `ice_density` and `gutter_length`) are done through a config file passed to `compass setup` with `-f`. 
```
[parallel]
account = t24_aishydro

[job]
qos = debug

wall_time = 0:10:00

# config options for the mesh (i.e. initial condition) setup
[mesh]

# Nominal cell spacing (m). Resolution is "nominal" since the true cell spacing
# will be determined such that the cell center to cell center length of the
# entire domain in the y direction is exactly the required y domain length (80 km).
resolution = 8000.

# length (m) to extend the eastern domain boundary by. Needed for simulations
# that use a dynamic calving law, where the calving front will be irregularly
# shaped. Any value less than `2*resolution` will be ignored, as the default
# gutter length is 2 gridcells.
gutter_length = 0.

# ice density (kg m^{-3}). MISMIP+ uses 918 (Table 1. Asay-Davis et al. 2016),
# but MALI defaults to 910. The user can choose if they want to stricly follow
# MISMIP+ or use the default MALI value.
ice_density = 918.
```
All variations of the test are setup with: 
```
compass setup -n 94 -w /lustre/scratch5/anolan/mismip+_pr_testing -f spin_up.cfg
```
And run by executing: 
```
cd /lustre/scratch5/anolan/mismip+_pr_testing/landice/mismipplus/spin_up
sbatch job_script.sh
```
Then once the test case is done, run: 
```
 mv *.log {resolution}m
 mv compass.o* {resolution}m
```
where `{resolution}` is the resolution of the testcase as specified in the config file (at the time of compass setup). 
The above command is need to prevent the log files being overwritten by successive tests at different resolutions. 
 
### (a): 1 km 
__Custom config options__: 
```
[mesh]
resolution = 1000.
```
__Output__: 
![Screenshot 2024-02-21 at 12 58 24 PM](https://github.com/MPAS-Dev/compass/assets/32367657/cd938020-9334-4822-84b4-0b35f6d529e3)

### (b):  4 km
__Custom config options__: 
```
[mesh]
resolution = 4000.
gutter_length = 12000.
ice_density = 218.
```
(_Note_: a density of 218. is totally arbitrary and just used to make sure the density in the namelist reflects the value requested in the config file). 

__Output__: 
![Screenshot 2024-02-21 at 1 00 00 PM](https://github.com/MPAS-Dev/compass/assets/32367657/1db347e6-8863-43df-a216-d8ba78737dab)
From`/lustre/scratch5/anolan/mismip+_pr_testing/landice/mismipplus/spin_up` running `less 4000m/run_model/namelist.landice` shows: 
![Screenshot 2024-02-21 at 1 00 43 PM](https://github.com/MPAS-Dev/compass/assets/32367657/238707ef-1246-40ea-b6c8-ab7c37b05075)
where density in the namelist properly reflects what's in the config file. 

### (c) 8000 m 
__Custom config options__: 
```
[mesh]
resolution = 1000.
ice_density = 910.
```
__Output__: 
![Screenshot 2024-02-21 at 1 04 14 PM](https://github.com/MPAS-Dev/compass/assets/32367657/8837cc62-1062-4265-83ff-42f478a3f7d0)
From`/lustre/scratch5/anolan/mismip+_pr_testing/landice/mismipplus/spin_up` running `less 8000m/run_model/namelist.landice` shows: 
![Screenshot 2024-02-21 at 1 05 09 PM](https://github.com/MPAS-Dev/compass/assets/32367657/e46359ec-5260-400f-a23c-c30ff585598f)
where density in the namelist properly reflects what's in the config file.

### (d) 16000 m 
 __Custom config options__: 
```
[mesh]
resolution = 16000.
```
__Output__: 
![Screenshot 2024-02-21 at 1 06 31 PM](https://github.com/MPAS-Dev/compass/assets/32367657/4ad42953-378b-41a0-abd2-038a4be56839)

### Summary 
All variations of the `SpinUp` testcase generate the mesh/initial condition successfully and successfully execute the landice model until the wall clock time (10 minutes) is reached.

trhille commented on compass/landice/tests/mismipplus/spin_up/namelist.spin_up:
```diff
@@ -0,0 +1,4 @@
+config_stop_time = '20000-01-01_00:00:00'
```
```suggestion
config_stop_time = '20001-01-01_00:00:00'
```

trhille reviewed (approved):
@andrewdnolan, amazing work! I have two tiny comments, but this is ready to go.

matthewhoffman reviewed (approved):
@andrewdnolan , thanks for all the revisions on this PR!  I'm happy to merge it.  Before I do that, can you update the PR description describing the final product?  You're also welcome to make the small changes Trevor suggested if you like.  Once you're happy with things, let me know and I will merge.

andrewdnolan commented:
@trhille @matthewhoffman Thanks for all the help reviewing / work shopping this PR! 

I've updated the PR description to reflect the final product. I've also implemented those two requested changes from Trevor. 

So, I'm happy with where things are if you guys are.
